viz/RobotModel.cpp

```diff
@@ -399,21 +619,172 @@ osg::ref_ptr<osg::Node> RobotModel::makeOsg( boost::shared_ptr<urdf::ModelInterf
     return root_;
 }
 
+osg::Node* RobotModel::makeOsg( sdf::ElementPtr sdf_model )
+{
+    std::string model_name = sdf_model->GetAttribute("name")->GetAsString();
+    SdfElementPtrMap links = loadSdfModelLinks(sdf_model);
+
+    KDL::Tree tree;
+    kdl_parser::treeFromSdfModel(sdf_model, tree);
+
+    KDL::SegmentMap::const_iterator root = tree.getRootSegment();
+
+    std::vector<KDL::SegmentMap::const_iterator> segment_buffer;
+    segment_buffer.insert(segment_buffer.end(), root->second.children.begin(), root->second.children.end());
+
+    std::vector<osg::Node*> hook_buffer;
+    for(uint i = 0; i < root->second.children.size(); i++){
+        hook_buffer.push_back(original_root_);
+    }
+
+    original_root_name_ = root->second.segment.getName();
+
+    osg::Node* hook = 0;
+    while (!segment_buffer.empty()){
+
+        KDL::SegmentMap::const_iterator it = segment_buffer.back();
+        KDL::SegmentMap::const_iterator parent_it = it->second.parent;
+
+        segment_buffer.pop_back();
+        KDL::Segment kdl_segment = it->second.segment;
+        KDL::Segment kdl_parent_segment = parent_it->second.segment;
+
+        if (it->second.children.size() > 0){
+            segment_buffer.insert(segment_buffer.end(), it->second.children.begin(), it->second.children.end());
+        }
+
+        std::map<std::string, sdf::ElementPtr>::iterator link_itr = links.find(kdl_segment.getName());
+        std::map<std::string, sdf::ElementPtr>::iterator link_parent_itr = links.find(kdl_parent_segment.getName());
+
+        if (link_itr != links.end()){
```

Comment:
.... If this fails, it means you have a bug. Throw an exception !

~~~
if (link_itr == links.end())
  throw std::runtime_error("Internal error: expected to find a SDF link called " + kdl_segment.getName());
else if (link_parent_itr == links.end())
  throw std::runtime_error("Internal error: expected to find a SDF link called " + kdl_parent_segment.getName());
~~~